ote_cli/ote_cli/utils/hpo.py

```diff
@@ -0,0 +1,660 @@
+            hpopt_arguments["reduction_factor"] = hpopt_cfg.get("reduction_factor")
+            hpopt_arguments["num_trials"] = hpopt_cfg.get("num_trials")
+            hpopt_arguments["num_workers"] = (
+                num_available_gpus // self.num_gpus_per_trial
+            )
+
+        HpoManager.remove_empty_keys(hpopt_arguments)
+
+        self.hpo = hpopt.create(**hpopt_arguments)
+
+    def check_resumable(self):
+        """
+        Check if HPO could be resumed from the previous result.
+        If previous results are found, ask the user if resume or start from scratch.
+        """
+        resume_flag = False
+
+        hpo_previous_status = hpopt.get_previous_status(self.work_dir)
+
+        if hpo_previous_status in [
+            hpopt.Status.PARTIALRESULT,
+            hpopt.Status.COMPLETERESULT,
+        ]:
+            print(f"Previous HPO results are found in {self.work_dir}.")
+            retry_count = 0
+
+            while True:
+                if hpo_previous_status == hpopt.Status.PARTIALRESULT:
+                    user_input = input(
+                        "Do you want to resume HPO? [\033[1mY\033[0m/n]: "
+                    )
+                else:
+                    user_input = input(
+                        "Do you want to reuse previously found "
+                        "hyper-parameters? [\033[1mY\033[0m/n]: "
+                    )
+
+                input_len = len(user_input)
+
+                if input_len < 1:
+                    # It means that an user accepted the default choice.
+                    resume_flag = True
+                    break
+
+                if input_len == 1:
+                    user_input = user_input.lower()
+                    if user_input in ["y", "n"]:
+                        if user_input == "y":
+                            resume_flag = True
+                        break
+
+                retry_count += 1
+                if retry_count >= 5:
+                    print(
+                        "Your inputs are invalid. "
+                        "The whole program is terminating..."
+                    )
+                    sys.exit()
+
+                print("You should type 'y' or 'n'. Nothing means 'y'.")
+        return resume_flag
+
+    def run(self):
+        """Execute HPO according to configuration"""
+
+        proc_list = []
+        gpu_alloc_list = []
+        num_workers = 1
+        if self.algo == "asha":
+            num_workers = torch.cuda.device_count() // self.num_gpus_per_trial
+
+        while True:
+            num_active_workers = 0
+            for proc, gpu in zip(reversed(proc_list), reversed(gpu_alloc_list)):
+                if proc.is_alive():
+                    num_active_workers += 1
+                else:
+                    proc.close()
+                    proc_list.remove(proc)
+                    gpu_alloc_list.remove(gpu)
+
+            if num_active_workers == num_workers:
+                time.sleep(10)
```

Comment:
what is this for?

Reply:
This is to wait without using the CPU in a situation where there are currently no allocable resources. It is possible to reduce the waiting time a bit, but since the running time per HPO trial is relatively long, it was set to 10 seconds.